src/core/net/ip6.cpp

```diff
@@ -414,7 +414,7 @@ ThreadError Ip6::HandleDatagram(Message &message, Netif *netif, uint8_t interfac
 
 exit:
 
-    if (error == kThreadError_Drop)
+    if ((error == kThreadError_Drop) || (error == kThreadError_NoRoute))
```

Comment:
1, Sorry that I read an older version of the code. You have removed the returned error value.
2, I think that it is ok to always use kThreadError_Drop to free buffers here so far.